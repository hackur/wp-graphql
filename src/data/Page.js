import { toGlobalId } from 'graphql-relay';

const path = process.env.WP_PAGES_ENDPOINT || 'wp/v2/pages';

class Page {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return path;
  }
}

export default Page;
