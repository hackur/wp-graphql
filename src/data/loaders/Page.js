import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Page from 'data/Page';

const endpoint = Page.getEndpoint();

// Dataloader expects IDs that can be read by the REST API

export default function getPageLoaders() {
  let slugLoader;
  const pageLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, orderby: 'include', per_page: 100 },
    }).then(({ data: { body: pages } }) => {
      pages.forEach(page => {
        slugLoader.prime(page.slug, page);
      });
      return pages;
    })
  );
  slugLoader = new Dataloader(slugs =>
    fetchData(endpoint, { qs: { slug: slugs } })
      .then(({ data: { body } }) => body)
      // the REST API does not order by FIELD(slug, ....) yet
      .then(pages => slugs.map(slug => pages.find(page => slug === page.slug)))
      .then(pages => {
        pages.forEach(page => {
          pageLoader.prime(page.id, page);
        });
        return pages;
      })
  );
  return {
    load: async id => {
      const data = await pageLoader.load(id);
      return data ? Object.assign(new Page(), data) : null;
    },
    loadBySlug: async slug => {
      const data = await slugLoader.load(slug);
      return data ? Object.assign(new Page(), data) : null;
    },
  };
}
