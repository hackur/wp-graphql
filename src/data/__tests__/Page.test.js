import Page from 'data/Page';

describe('Test Page data access', () => {
  test('Get endpoint', () => {
    expect(Page.getEndpoint()).toMatchSnapshot();
  });
});
