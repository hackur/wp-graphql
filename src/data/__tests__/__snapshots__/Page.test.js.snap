// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Page data access Get endpoint 1`] = `"wp/v2/pages"`;
